asmcomp/cmmgen.ml

```diff
@@ -915,23 +915,23 @@ and transl_prim_2 env p arg1 arg2 dbg =
       let a2 = transl_unbox_int dbg env bi arg2 in
       mk_compare_ints dbg a1 a2
   | Pcompare_floats ->
-      let a1 = transl_unbox_float dbg env arg1 in
-      let a2 = transl_unbox_float dbg env arg2 in
-      let op1 = Cop(Ccmpf(CFgt), [a1; a2], dbg) in
-      let op2 = Cop(Ccmpf(CFlt), [a1; a2], dbg) in
-      let op3 = Cop(Ccmpf(CFeq), [a1; a1], dbg) in
-      let op4 = Cop(Ccmpf(CFeq), [a2; a2], dbg) in
-      (* If both operands a1 and a2 are not NaN, then op3 = op4 = 1,
-         and the result is op1 - op2.
-         If at least one of the operands is NaN,
-         then op1 = op2 = 0, and the result is op3 - op4,
-         which orders NaN before other values.
-         To detect if the operand is NaN, we use the property:
-         for all x, NaN is not equal to x, even if x is NaN.
-         Therefore, op3 is 0 if and only if a1 is NaN,
-         and op4 is 0 if and only if a2 is NaN.
-         See also caml_float_compare_unboxed in runtime/floats.c  *)
-      tag_int (add_int (sub_int op1 op2 dbg) (sub_int op3 op4 dbg) dbg) dbg
+      bind "float_cmp" (transl_unbox_float dbg env arg1) (fun a1 ->
+      bind "float_cmp" (transl_unbox_float dbg env arg2) (fun a2 ->
+        let op1 = Cop(Ccmpf(CFgt), [a1; a2], dbg) in
+        let op2 = Cop(Ccmpf(CFlt), [a1; a2], dbg) in
+        let op3 = Cop(Ccmpf(CFeq), [a1; a1], dbg) in
+        let op4 = Cop(Ccmpf(CFeq), [a2; a2], dbg) in
+        (* If both operands a1 and a2 are not NaN, then op3 = op4 = 1,
+           and the result is op1 - op2.
+           If at least one of the operands is NaN,
+           then op1 = op2 = 0, and the result is op3 - op4,
+           which orders NaN before other values.
+           To detect if the operand is NaN, we use the property:
+           for all x, NaN is not equal to x, even if x is NaN.
+           Therefore, op3 is 0 if and only if a1 is NaN,
+           and op4 is 0 if and only if a2 is NaN.
+           See also caml_float_compare_unboxed in runtime/floats.c  *)
+        tag_int (add_int (sub_int op1 op2 dbg) (sub_int op3 op4 dbg) dbg) dbg))
```

Comment:
Could this big and subtle block be moved to an auxiliary function, here or in `cmm_helpers`? It would be more natural with the rest of the function which has dedicated helpers in all cases.

Reply:
done.